x-pack/legacy/plugins/index_management/public/components/template_form/template_form.tsx

```diff
@@ -0,0 +1,245 @@
+import {
+  validateLogistics,
+  validateSettings,
+  validateMappings,
+  validateAliases,
+  TemplateValidation,
+} from '../../services/template_validation';
+import { StepProps } from './types';
+import { BASE_PATH } from '../../../common/constants';
+import { SectionError } from '..';
+
+interface Props {
+  onSave: (template: Template) => void;
+  clearSaveError: () => void;
+  isSaving: boolean;
+  saveError: any;
+  template: Template;
+  isEditing?: boolean;
+}
+
+const defaultValidation = {
+  isValid: true,
+  errors: {},
+};
+
+const stepComponentMap: { [key: number]: React.FunctionComponent<StepProps> } = {
+  1: StepLogistics,
+  2: StepSettings,
+  3: StepMappings,
+  4: StepAliases,
+  5: StepReview,
+};
+
+const stepValidationMap: { [key: number]: any } = {
+  1: validateLogistics,
+  2: validateSettings,
+  3: validateMappings,
+  4: validateAliases,
+};
+
+export const TemplateForm: React.FunctionComponent<Props> = ({
+  template: initialTemplate,
+  onSave,
+  isSaving,
+  saveError,
+  clearSaveError,
+  isEditing,
+}) => {
+  // hooks
+  const [currentStep, setCurrentStep] = useState<number>(1);
+  const [maxCompletedStep, setMaxCompletedStep] = useState<number>(0);
+  const [template, setTemplate] = useState<Template>(initialTemplate);
+  const [validation, setValidation] = useState<{ [key: number]: TemplateValidation }>({
+    1: defaultValidation,
+    2: defaultValidation,
+    3: defaultValidation,
+    4: defaultValidation,
+    5: defaultValidation,
+  });
+
+  const lastStep = Object.keys(stepComponentMap).length;
+
+  const CurrentStepComponent = stepComponentMap[currentStep];
+
+  const validateStep = stepValidationMap[currentStep];
+
+  const stepErrors = validation[currentStep].errors;
+  const isStepValid = validation[currentStep].isValid;
+
+  const validationErrors = Object.keys(stepErrors).reduce((acc: any, key: string) => {
+    return [...acc, ...stepErrors[key]];
+  }, []);
+
+  const updateTemplate = (updatedTemplate: Partial<Template>): void => {
+    const newTemplate = { ...template, ...updatedTemplate };
+
+    setTemplate(newTemplate);
+  };
+
+  const updateCurrentStep = (nextStep: number, nextMaxCompletedStep: number) => {
+    // All steps needs validation, except for the last step
+    const shouldValidate = currentStep !== lastStep;
+
+    if (shouldValidate) {
+      const stepValidation = validateStep(template);
+
+      const { isValid: isCurrentStepValid } = stepValidation;
+
+      const newValidation = {
+        ...validation,
+        ...{
+          [currentStep]: stepValidation,
+        },
+      };
+
+      setValidation(newValidation);
+
+      // If step is invalid do not let user proceed
+      if (!isCurrentStepValid) {
+        return;
+      }
+    }
+
+    setCurrentStep(nextStep);
+    setMaxCompletedStep(nextMaxCompletedStep);
+    clearSaveError();
+  };
+
+  const onBack = () => {
+    const prevStep = currentStep - 1;
+    updateCurrentStep(prevStep, prevStep - 1);
```

Comment:
The `maxCompletedStep` should not go down (always up). What was not working about my previous example?